aggregator/src/aggregator/aggregation_job_writer.rs

```diff
@@ -344,25 +344,18 @@ impl<const SEED_SIZE: usize, Q: CollectableQueryType, A: vdaf::Aggregator<SEED_S
                     } else if agg_job_op == &Operation::Update
                         && !matches!(agg_job.state(), AggregationJobState::InProgress)
                     {
-                        // GC hack: if we are Putting the batch, that means that it does not exist
-                        // in the datastore. But since we first write the batch when an aggregation
-                        // job referencing that batch is created, and we are completing the
-                        // aggregation job here, we must have deleted the batch at some point
-                        // between creating & completing this aggregation job. This should only be
-                        // possible if the batch is GC'ed, as part of a time-interval task. In that
-                        // case, it is acceptable to skip writing the batch entirely; and indeed, we
-                        // must do so, since otherwise we might underflow the
-                        // outstanding_aggregation_jobs counter.
+                        // If we are Putting the batch, that means that it does not exist in the
+                        // datastore. But since we first write the batch when an aggregation job
+                        // referencing that batch is created, and we are completing the aggregation
+                        // job here, we must have deleted the batch at some point between creating &
+                        // completing this aggregation job. This should only be possible if the
+                        // batch is GC'ed. In that case, it is acceptable to skip writing the batch
+                        // entirely; and indeed, we must do so, since otherwise we might underflow
+                        // the outstanding_aggregation_jobs counter.
                         //
                         // See https://github.com/divviup/janus/issues/2464 for more detail.
                         if batch_op == Operation::Put {
-                            // Guard to ensure we are in the situation we think we're in: we are in
-                            // a time-interval task, and the batch interval is past the GC window.
-                            if !matches!(self.task.query_type(), QueryType::TimeInterval)
-                                || !Q::to_batch_interval(batch_identifier)
-                                    .map(|interval| interval.end() < tx.clock().now())
-                                    .unwrap_or(false)
-                            {
+                            if Q::is_batch_garbage_collected(tx.clock(), batch_identifier) != Some(true) {
```

Comment:
Hmm, there is a new functional change here--a fixed size task in the same situation would go down the panic, where it didn't before. We're no longer assuming that a fixed size batch is GC'd.

I'm not sure that we want this, it seems possible to run into this state with a fixed size batch being GC'd, where a panic seems inappropriate. Thoughts?

Reply:
I think this is not a logic change. The old logic to trigger a panic was:

```
!matches!(self.task.query_type(), QueryType::TimeInterval)
    || !Q::to_batch_interval(batch_identifier)
        .map(|interval| interval.end() < tx.clock().now())
        .unwrap_or(false)
```

...which would panic if the query type was `FixedSize`, or if the query type was `TimeInterval` and the batch interval's end was not before now.

The new logic panics if `Q::is_batch_garbage_collected` returns something other than `Some(true)`. This will happen if the query type is `FixedSize` (since `FixedSize`'s implementation always returns `None`), or if the query type is `TimeInterval` and the batch interval's end is not before now.

(please double-check this, I've looked real hard at it a couple of times but maybe I'm missing something obvious)

Reply:
Agreed, LGTM. Tangentially, I'll note that `!matches!(..., ...)` is easy to misread, I caught a bug elsewhere a while back from a related typo.

Reply:
Ah yep with fresh eyes I see that you're right. Sorry for the misdirect!